File: Arrays & Strings/One Away.py
class Solution:

    def replaceOneLetter(self, str1, str2):

        # OBJECTIVE: Traverse strings and flag when one edit is made

        # Change string to a list
        str1_list = list(str1)
        str2_list = list(str2)

        # Traverse string
        madeAnEdit = False
        for idx in range(len(str1_list)):

            # If both letters don't match
            if str1_list[idx] != str2_list[idx]:

                # If an edit has already been made, return False
                if madeAnEdit:
                    return False

                # If not, update boolean variable
                madeAnEdit = True

        # If the for-loop ended, then both strings are only different by 1 edit
        return True

    def deleteOrAdd(self, smallStr, bigStr):

        # OBJECTIVE: Traverse both strings and flag when a letter needs to be added or deleted from either strings

        # Change string to a list
        smallStr_List = list(smallStr)
        bigStr_List = list(bigStr)

        # Traverse both lists
        smallIdx = 0
        bigIdx = 0
        madeAnEdit = False
        while smallIdx < len(smallStr_List) and bigIdx < len(bigStr_List):

            # If both letters don't match
            if smallStr_List[smallIdx] != bigStr_List[bigIdx]:

                # If a change has already been made, then return false
                if madeAnEdit:
                    return False

                # If this is the first edit, move bigIdx by 1 and update boolean variable
                # E.g., ple & pale
                bigIdx += 1
                madeAnEdit = True

                continue

            smallIdx += 1
            bigIdx += 1

        return True

    def solOne(self, str1, str2):
        
        # If both strings are different in length by 2, then return false
        # E.g., bake & bakery
        if abs(len(str1) - len(str2)) >= 2: # <= Add abs() because wasn't sure which stirng is bigger
            return False

        # If both strings share the same length, check if one letter needs to be change in either strings
        if len(str1) == len(str2):
            return self.replaceOneLetter(str1, str2)

        # If str1 is bigger by 1 character
        elif len(str1) > len(str2):
            return self.deleteOrAdd(str2, str1)

        # If str2 is shorter by 1 character
        elif len(str1) < len(str2):
            return self.deleteOrAdd(str1, str2)

sol = Solution()
print(sol.solOne("pale", "ple"))
print(sol.solOne("pales", "pale"))
print(sol.solOne("pale", "bale"))
print(sol.solOne("pale", "bake"))

File: Arrays & Strings/One Away.java
class OneAway {

    public static boolean replaceOneLetter(String str1, String str2) {

        // OBJECTIVE: Traverse strings and flag when one edit is made

        // Change both strings into a char array
        char[] str1_array = str1.toCharArray();
        char[] str2_array = str2.toCharArray();

        // Traverse array
        boolean madeAnEdit = false;
        for (int i=0; i<str1_array.length; i++) {

            // If both letters don't match
            if (str1_array[i] != str2_array[i]) {

                // If an edit has already been made, return false
                // NOTE: At most 1 edit can be made
                if (madeAnEdit) {
                    return false;
                }

                // If not, set variable to true
                madeAnEdit = true;
            }
        }

        // If for-loop ended, return true
        return true;
    }

    public static boolean deleteOrAdd(String smallStr, String bigStr) {

        // OBJECTIVE: Traverse both strings and flag when a letter needs to be added or deleted from either strings

        // Change both strings to an array
        char[] smallArray = smallStr.toCharArray();
        char[] bigArray = bigStr.toCharArray();

        // Traverse both arrays
        int smallIdx = 0;
        int bigIdx = 0;
        boolean madeAnEdit = false;

        while (smallIdx < smallArray.length && bigIdx < bigArray.length) {

            // If both letters don't match
            if (smallArray[smallIdx] != bigArray[bigIdx]) {

                // If an edit has already been made, return false
                if (madeAnEdit) {return false;}

                // If not, move bigIdx by 1 and update boolean variable
                // E.g., ple & pale
                bigIdx++;
                madeAnEdit = true;

                continue;
            }

            // Update pointers
            smallIdx++;
            bigIdx++;
        }

        // If while-loop ended, return true
        return true;
    }

    public static boolean solOne(String str1, String str2) {
        
        /*
            OBJECTIVE: Check if both strings are the same with at-most 1 edit permitted (replace, drop, or insert a letter)
            
            Time complexity: O(n) where n = length of either string if replaceOneLetter() is executed. If deleteOrAdd() is
                            executed instead, then O(b) where b = length of bigger string because bigIdx can move 1 index more
                            than smallIdx if a mismatched letter is found
            
            Space complexity: O(n) where n = length of either string if replaceOneLetter() is executed because 2 new lists are
                            created. In deleteOrAdd(), it will be O(S + B) where S = length of small string and B = length of
                            big string because 2 new lists are created for each string.
        */

        // If both strings are different in length by 2, return false
        if (Math.abs(str1.length() - str2.length()) >= 2) {
            return false;
        }

        // If both strings share the same length, check if 1 letter needs to be changed
        if (str1.length() == str2.length()) {
            return replaceOneLetter(str1, str2);
        }

        // If str1 is bigger by 1 character
        else if (str1.length() > str2.length()) {
            return deleteOrAdd(str2, str1);
        }

        // If str1 is smaller by 1 character
        else {
            return deleteOrAdd(str1, str2);
        }
    }

    public static void main(String args[]) {

        System.out.println(solOne("pale", "ple"));
        System.out.println(solOne("pales", "pale"));
        System.out.println(solOne("pale", "bale"));
        System.out.println(solOne("pale", "bake"));
    }
}